packages/core/src/ckb/epoch.ts

```diff
@@ -0,0 +1,296 @@
+    }
+  | [NumLike, NumLike, NumLike];
+
+@mol.codec(
+  mol
+    .struct({
+      length: mol.uint(2, true),
+      index: mol.uint(2, true),
+      number: mol.uint(3, true),
+    })
+    .mapIn((encodable: EpochLike) => Epoch.from(encodable)),
+)
+/**
+ * Epoch
+ *
+ * Represents a timestamp-like epoch as a mixed whole number and fractional part:
+ * - number: whole units
+ * - index: numerator of the fractional part
+ * - length: denominator of the fractional part (must be > 0)
+ *
+ * The fractional portion is index/length. Instances normalize fractions where
+ * appropriate (e.g., reduce by GCD, carry whole units).
+ */
+export class Epoch extends mol.Entity.Base<EpochLike, Epoch>() {
+  /**
+   * Construct a new Epoch.
+   *
+   * The constructor enforces a positive `length` (denominator). If `length`
+   * is non-positive an Error is thrown.
+   *
+   * @param number - Whole number portion of the epoch.
+   * @param index - Fractional numerator.
+   * @param length - Fractional denominator (must be > 0).
+   */
+  public constructor(
+    public readonly number: Num,
+    public readonly index: Num,
+    public readonly length: Num,
+  ) {
+    // Ensure the epoch has a positive denominator.
+    if (length <= Zero) {
+      throw new Error("Non positive Epoch length");
+    }
+    super();
+  }
+
+  /**
+   * @deprecated use `number` instead
+   * Backwards-compatible array-style index 0 referencing the whole epoch number.
+   */
+  get 0(): Num {
+    return this.number;
+  }
+
+  /**
+   * @deprecated use `index` instead
+   * Backwards-compatible array-style index 1 referencing the epoch fractional numerator.
+   */
+  get 1(): Num {
+    return this.index;
+  }
+
+  /**
+   * @deprecated use `length` instead
+   * Backwards-compatible array-style index 2 referencing the epoch fractional denominator.
+   */
+  get 2(): Num {
+    return this.length;
+  }
+
+  /**
+   * Create an Epoch from an EpochLike value.
+   *
+   * Accepts:
+   * - an Epoch instance (returned as-is)
+   * - an object { number, index, length } where each field is NumLike
+   * - a tuple [number, index, length] where each element is NumLike
+   *
+   * All returned fields are converted to `Num` using `numFrom`.
+   *
+   * @param epochLike - Value to convert into an Epoch.
+   * @returns A new or existing Epoch instance.
+   */
+  static from(epochLike: EpochLike): Epoch {
+    if (epochLike instanceof Epoch) {
+      return epochLike;
+    }
+
+    let number: NumLike, index: NumLike, length: NumLike;
+    if (epochLike instanceof Array) {
+      [number, index, length] = epochLike;
+    } else {
+      ({ number, index, length } = epochLike);
+    }
+
+    return new Epoch(numFrom(number), numFrom(index), numFrom(length));
+  }
+
+  /**
+   * Return an epoch representing zero (0 + 0/1).
+   */
+  static zero(): Epoch {
+    return new Epoch(0n, 0n, numFrom(1));
+  }
+
+  /**
+   * Return an epoch representing one (1 + 0/1).
+   */
+  static one(): Epoch {
```

Comment:
Hmm... when will we need this? It's obviously not CKB's epoch one.

Reply:
Done, removed :+1: